pkg/manifest/manifest_loader.go

```diff
@@ -307,20 +325,46 @@ func toEnvironments(context *ManifestLoaderContext, groups []group) (map[string]
 
 		groupNames[group.Name] = true
 
-		for _, conf := range group.Environments {
-			env, configErrors := toEnvironment(context, conf, group.Name)
+		for j, env := range group.Environments {
 
-			if configErrors != nil {
-				errors = append(errors, configErrors...)
+			if env.Name == "" {
+				errors = append(errors, newManifestLoaderError(context.ManifestPath, fmt.Sprintf("missing environment name in group %q on index `%d`", group.Name, j)))
+				continue
+			}
+
+			if envNames[env.Name] {
+				errors = append(errors, newManifestLoaderError(context.ManifestPath, fmt.Sprintf("duplicated environment name %q", env.Name)))
 				continue
 			}
+			envNames[env.Name] = true
 
-			if _, found := environments[env.Name]; found {
-				errors = append(errors, newManifestLoaderError(context.ManifestPath, fmt.Sprintf("duplicated environment name `%s`", env.Name)))
+			// skip loading if environments is not empty, the environments does not contain the env name, or the group should not be included
+			if shouldSkipEnv(context, group, env) {
+				log.Debug("skipping loading of environment %q", env.Name)
 				continue
 			}
 
-			environments[env.Name] = env
+			parsedEnv, configErrors := parseEnvironment(context, env, group.Name)
+
+			if configErrors != nil {
+				errors = append(errors, configErrors...)
+				continue
+			}
+
+			environments[parsedEnv.Name] = parsedEnv
+		}
+	}
+
+	// validate that all required groups & environments are included
+	for _, g := range context.Groups {
+		if !groupNames[g] {
+			errors = append(errors, newManifestLoaderError(context.ManifestPath, fmt.Sprintf("required to load group %q, but could not found", g)))
+		}
+	}
+
+	for _, e := range context.Environments {
+		if !envNames[e] {
+			errors = append(errors, newManifestLoaderError(context.ManifestPath, fmt.Sprintf("required to load environment %q, but could not found", e)))
```

Comment:
```suggestion
			errors = append(errors, newManifestLoaderError(context.ManifestPath, fmt.Sprintf("requested environment %q not found", e)))
```

alternative language fix: "but could not find it" - but I think these messages can be shorter and still say what's wrong clearly